nixos/kea: init

mweinelt opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

- [x] Adds module for the Kea DHCP servers. 
- [x] Updates the kea prometheus-exporter module to be compatible with the kea module.
- [x] Update to the prometheus-exporters default serviceConfig, so a specific username can be set, and still be overridden by endusers.
- [x] Adopt maintainership of kea
- [x] NixOS test for kea
- [x] Changelog entry

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mweinelt committed:
nixos/kea: init

mweinelt committed:
nixos/prometheus: allow overriding default export user

mweinelt committed:
nixos/tests/prometheus-exporters/kea: migrate to kea nixos module

mweinelt committed:
kea: adopt maintainership

mweinelt commented:
@GrahamcOfBorg test prometheus-exporters

ajs124 commented on nixos/modules/services/networking/kea.nix:
```diff
@@ -0,0 +1,363 @@
+  dhcp4Config = format.generate "kea-dhcp4.conf" {
+    Dhcp4 = cfg.dhcp4.settings;
+  };
+  dhcp6Config = format.generate "kea-dhcp4.conf" {
```
Might be confusing to call both these `kea-dhcp4.conf`.

mweinelt replied:
Is might copy/paste really that bad? :open_mouth:

ajs124 commented on nixos/modules/services/networking/kea.nix:
```diff
@@ -0,0 +1,363 @@
+    };
+  in mkIf cfg.enable (mkMerge [
+  {
+    environment.systemPackages = with pkgs; [ kea ];
```
Why not use `package` here?

mweinelt replied:
Because `package` was a last minute refactor and I missed this one. :man_facepalming:

mweinelt committed:
nixos/tests/kea: init

A simple DHCPv4 test case, with borrowed structure from the
systemd-networkd-dhcpserver test.

mweinelt committed:
kea: setup passthru test

mweinelt commented on nixos/modules/services/networking/kea.nix:
```diff
@@ -0,0 +1,363 @@
+      StateDirectory = "kea";
+      UMask = "0077";
+    };
+  in mkIf cfg.enable (mkMerge [
```
```suggestion
  in mkIf (cfg.ctrl-agent.enable or cfg.dhcp4.enable or cfg.dhcp6.enable or cfg.dhcp-ddns.enable) (mkMerge [
```

Maybe this would be better?

mweinelt committed:
nixos/doc: add release note entry for kea module

ajs124 commented on nixos/modules/services/networking/kea.nix:
```diff
@@ -0,0 +1,361 @@
+
+    environment.etc."kea/dhcp4-server.conf".source = dhcp4Config;
+
+    systemd.services.kea-dhcp4-server = mkIf cfg.dhcp4.enable {
```
This mkIf seems redundant

ajs124 commented on nixos/modules/services/networking/kea.nix:
```diff
@@ -0,0 +1,361 @@
+
+    environment.etc."kea/dhcp6-server.conf".source = dhcp6Config;
+
+    systemd.services.kea-dhcp6-server = mkIf cfg.dhcp6.enable {
```
Same here

ajs124 commented on nixos/modules/services/networking/kea.nix:
```diff
@@ -0,0 +1,361 @@
+
+    environment.etc."kea/dhcp-ddns.conf".source = dhcpDdnsConfig;
+
+    systemd.services.kea-dhcp-ddns-server = mkIf cfg.dhcp-ddns.enable {
```
And here

ajs124 reviewed (changes requested):
Some parts could probably be de-duplicated, although I'm not sure that would help the readability of the implementation

ajs124 commented on nixos/tests/kea.nix:
```diff
@@ -0,0 +1,73 @@
+
+    client = { config, pkgs, ... }: {
+      virtualisation.vlans = [ 1 ];
+      systemd.services.systemd-networkd.environment.SYSTEMD_LOG_LEVEL = "debug";
```
Why is this set?

mweinelt replied:
So you can see DHCP events happening on the client side.

ajs124 reviewed (approved):
Probably alright
